hibernate-reactive-core/src/main/java/org/hibernate/reactive/stage/impl/StageSessionFactoryImpl.java

```diff
@@ -78,23 +80,42 @@ public StageSessionFactoryImpl(SessionFactoryImpl delegate) {
 	CompletionStage<Stage.Session> newSession() {
 		SessionCreationOptions options = options();
 		return stage( v -> connection( options.getTenantIdentifier() )
-				.thenApply( connection -> new ReactiveSessionImpl( delegate, options, connection ) )
+				.thenCompose( connection -> create( connection, () -> new ReactiveSessionImpl( delegate, options, connection ) ) )
 				.thenApply( s -> new StageSessionImpl(s, this) ) );
 	}
 
 	CompletionStage<Stage.Session> newSession(String tenantId) {
 		return stage( v -> connection( tenantId )
-				.thenApply( connection -> new ReactiveSessionImpl( delegate, options( tenantId ), connection ) )
+				.thenCompose( connection -> create( connection, () -> new ReactiveSessionImpl( delegate, options( tenantId ), connection ) ) )
 				.thenApply( s -> new StageSessionImpl(s, this) ) );
 	}
 
 	CompletionStage<Stage.StatelessSession> newStatelessSession() {
 		SessionCreationOptions options = options();
 		return stage( v -> connection( options.getTenantIdentifier() )
-				.thenApply( connection -> new ReactiveStatelessSessionImpl( delegate, options, connection ) )
+				.thenCompose( connection -> create( connection, () -> new ReactiveStatelessSessionImpl( delegate, options, connection ) ) )
 				.thenApply( s -> new StageStatelessSessionImpl(s, this) ) );
 	}
 
+	/**
+	 * Close the connection if something goes wrong during the creation of the session
```

Comment:
What kind of things could go wrong?

Reply:
Multitenancy.
Check the constructor for `AbstractSharedSessionContract`. 

If we don't close the we are going to see the exception but the connection will never get closed, leading to the app getting stuck or throwing  TimeOut exceptions that are not easy to find.

Reply:
I mean, Multitenancy is the easier to see, but there are all sort of initialisation when we create the session. So, there might be something else

Reply:
Probably, a better solution would be to wait to get the connection until we are ready with the session.

Reply:
> Probably, a better solution would be to wait to get the connection until we are ready with the session.

Hrm so you're worried about closing the connection because in `withSession()` we don't use a `ProxyConnection`. I see.